Trigger a signal in the UI Process when the web page provides a spatial backdrop

jeanhbr opened the pull request:
#### 4cd69c9df3932e49035491d1da0789ed0a7a22a5
<pre>
Trigger a signal in the UI Process when the web page provides a spatial backdrop
<a href="https://bugs.webkit.org/show_bug.cgi?id=286510">https://bugs.webkit.org/show_bug.cgi?id=286510</a>
<a href="https://rdar.apple.com/141175983">rdar://141175983</a>

Reviewed by Ada Chan and Chris Dumez.

Expose the value of a spatial backdrop link tag to the UI Process.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::spatialBackdropSource):
(WebCore::Document::spatialBackdropLinkElementChanged):
(WebCore::Document::determineActiveSpatialBackdropSource):
(WebCore::Document::spatialBackdropSourceChanged):
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/SpatialBackdropSource.cpp: Added.
(WebCore::SpatialBackdropSource::SpatialBackdropSource):
* Source/WebCore/dom/SpatialBackdropSource.h: Added.
(WebCore::SpatialBackdropSource::SpatialBackdropSource):
(WebCore::operator==):
Transports the spatial backdrop properties that are useful to know for the UI
Process, this includes the source (current URL) from which the backdrop was
provided, and the URL of both the provided model and environmentmap.

* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::attributeChanged):
(WebCore::HTMLLinkElement::process):
(WebCore::HTMLLinkElement::removedFromAncestor):
(WebCore::HTMLLinkElement::isURLAttribute const):
(WebCore::HTMLLinkElement::environmentmap const):
(WebCore::HTMLLinkElement::isSpatialBackdrop const):

* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLLinkElement.idl:
* Source/WebCore/html/LinkRelAttribute.cpp:
* Source/WebCore/html/LinkRelAttribute.h:
Add the &quot;spatial-backdrop&quot; rel, and the &quot;environmentmap&quot; attribute to
the link tag.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::spatialBackdropSourceChanged const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::spatialBackdropSource const):
* Source/WebCore/page/Page.h:

* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _spatialBackdropSource]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.mm: Added.
(-[_WKSpatialBackdropSource initWithSpatialBackdropSource:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSourceInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::spatialBackdropSourceWillChange):
(WebKit::PageClientImplCocoa::spatialBackdropSourceDidChange):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::spatialBackdropSourceWillChange):
(WebKit::PageClient::spatialBackdropSourceDidChange):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::spatialBackdropSource const):
(WebKit::WebPageProxy::spatialBackdropSourceChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::spatialBackdropSourceChanged const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didFinishLoad):
(WebKit::WebPage::spatialBackdropSourceChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/289748@main">https://commits.webkit.org/289748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4572fe95ad516009a57e1da36080f761d8cba338

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87875 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7391 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42264 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92769 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/38625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89926 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7772 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15566 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/38625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90877 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/5956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79513 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/48216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/5741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/33922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37732 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/80673 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/34800 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94639 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86650 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15043 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15298 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75369 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/75933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18669 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/20317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/8047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15061 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109144 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14803 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26245 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18248 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/16585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebCore/dom/Document.h:
```diff
@@ -2219,6 +2235,10 @@ class Document
     Color m_applicationManifestThemeColor;
 
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+    SpatialBackdropSource m_cachedSpatialBackdropSource;
```
Curious if there's a reason to not make this an optional type

jeanhbr replied:
I think making this optional would indeed be better, thanks! 🙏🏼

achan00 commented on Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.h:
```diff
@@ -0,0 +1,36 @@
+WK_CLASS_AVAILABLE(visionos(WK_XROS_TBA))
+@interface _WKSpatialBackdropSource : NSObject
+
+@property (nonatomic, readonly, nonnull) NSURL *sourceURL WK_API_AVAILABLE(visionos(WK_XROS_TBA));
```
Wrap the interface with `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END` and then the `nonnull` attribute is not needed anymore here.

achan00 commented on Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.mm:
```diff
@@ -0,0 +1,53 @@
+@implementation _WKSpatialBackdropSource
+
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+- (instancetype _Nullable)initWithSpatialBackdropSource:(const WebCore::SpatialBackdropSource&) spatialBackdropSource
```
Nit: we don't need `_Nullable` in the implementation. We only need this in the header.

achan00 commented on Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.mm:
```diff
@@ -0,0 +1,53 @@
+        return nil;
+
+    _sourceURL = [spatialBackdropSource.sourceURL copy];
+    if (spatialBackdropSource.modelURL.has_value())
```
Nit: is `modelURL` an `std::optional`? If so, can just check it as bool directly.

achan00 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
```diff
@@ -396,6 +396,10 @@ struct PerWebProcessState {
 #endif
 
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+    _WKSpatialBackdropSource *_cachedSpatialBackdropSource;
```
We should use a `RetainPtr<_WKSpatialBackdropSource>` for this.

achan00 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -4705,6 +4707,17 @@ - (void)_setCanUseCredentialStorage:(BOOL)canUseCredentialStorage
+{
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+    if (!_cachedSpatialBackdropSource)
+        _cachedSpatialBackdropSource = [[_WKSpatialBackdropSource alloc] initWithSpatialBackdropSource:_page->spatialBackdropSource()];
```
And once `_cachedSpatialBackdropSource` is a `RetainPtr`, then this should use `adoptNS(...)` here.

jeanhbr commented on Source/WebCore/html/LinkRelAttribute.cpp:
```diff
@@ -65,6 +65,9 @@ static constexpr std::pair<ComparableLettersLiteral, LinkTypeDetails> linkTypesA
     { "prefetch"_s, { [](auto document) { return document.settings().linkPrefetchEnabled(); }, [](auto relAttribute) { relAttribute.isLinkPrefetch = true; } } },
     { "preload"_s, { [](auto document) { return document.settings().linkPreloadEnabled(); }, [](auto relAttribute) { relAttribute.isLinkPreload = true; } } },
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+    { "spatial-backdrop"_s, { [](auto) { return true; }, [](auto relAttribute) { relAttribute.isSpatialBackdrop = true; } } },
```
Need to make it conditional on the settings I've added.

annevk commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -202,6 +205,10 @@ void HTMLLinkElement::attributeChanged(const QualifiedName& name, const AtomStri
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+        if (equalLettersIgnoringASCIICase(oldValue, "spatial-backdrop"_s) && !equalLettersIgnoringASCIICase(newValue, "spatial-backdrop"_s))
+            document().spatialBackdropLinkElementChanged();
+#endif
```
This seems bad. `m_relAttribute` and `parsedRel` already store the relevant booleans for this.

This also wouldn't catch mutations that differ by whitespace.

annevk replied:
It also seems like you end up invoking this multiple times in certain cases. Once through `process()` on the line before and once through this?

jeanhbr replied:
Thanks! Changed this to:
```
if (wasSpatialBackdrop && !m_relAttribute.isSpatialBackdrop)
   document().spatialBackdropLinkElementChanged();
```
Maybe I misunderstood your comment, but I don't think this would trigger `spatialBackdropLinkElementChanged` twice; since in `process()` we check `m_relAttribute.isSpatialBackdrop` before triggering `spatialBackdropLinkElementChanged`. In our case, when we end up with `(wasSpatialBackdrop && !m_relAttribute.isSpatialBackdrop) == true`, that would mean that `m_relAttribute.isSpatialBackdrop == false`, thus triggering only the call above and not the one in `process()`.

annevk commented on Source/WebCore/html/HTMLLinkElement.idl:
```diff
@@ -27,6 +27,7 @@
     [Reflect] attribute boolean disabled;
     [CEReactions=NotNeeded, Reflect] attribute DOMString charset;
     [CEReactions=NotNeeded, Reflect, URL] attribute USVString href;
+    [CEReactions=NotNeeded, Conditional=WEB_PAGE_SPATIAL_BACKDROP, EnabledBySetting=WebPageSpatialBackdropEnabled, Reflect, URL] attribute USVString environmentmap;
```
environmentMap

annevk commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3410,6 +3410,18 @@ Color Page::themeColor() const
+std::optional<SpatialBackdropSource> Page::spatialBackdropSource() const
+{
+    auto* localMainFrame = dynamicDowncast<LocalFrame>(mainFrame());
+    auto* document = localMainFrame ? localMainFrame->document() : nullptr;
```
New code should prolly use RefPtr/Ref.

annevk commented on Source/WebKit/UIProcess/API/Cocoa/WKPreferences.mm:
```diff
@@ -1713,6 +1713,24 @@ - (BOOL)_spatialVideoEnabled
+#else
+    return NO;
+#endif
+}
```
I kinda doubt we want to modify this file? As I understand it this is a legacy file from before UniviedWebPreferences.yaml.

jeanhbr replied:
Interesting, I'm not sure what's best to do here. I see recent commits to this file, adding new preferences. It looks like people still keep this one up to date?

annevk replied:
That's fair, but I think new things only get added when needed. Will there be a caller for this?

jeanhbr replied:
Yeah good point, no there won't be any caller for this so I'm removing it ✅

annevk commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1330,6 +1330,54 @@ const Color& Document::themeColor()
+
+        auto modelURL = linkElement.href();
+        if (!modelURL.isValid())
+            continue;
```
I think a standard for a feature like this would just pick the first qualifying `link` element and then if it doesn't work give up. Or it would actually fully try each of them. But not only validate a single condition but not any of the others.

jeanhbr replied:
Got it, so probably doing `return std::nullopt;` here instead of `continue;`?

annevk replied:
Attempting the load so we end up firing an `error` event would probably be more useful. (This would also be good to test.)

jeanhbr replied:
As we won't load the resource here, it won't fire an error automatically I believe. But we could maybe add an error in the console here if `!modelURL.isValid()`.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1330,6 +1330,54 @@ const Color& Document::themeColor()
 }
 
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+const std::optional<SpatialBackdropSource>& Document::spatialBackdropSource()
```
Ideally, this function would be marked as `const` and inlined in the header.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1330,6 +1330,54 @@ const Color& Document::themeColor()
+        return;
+
+    auto newSpatialBackdropSource = determineActiveSpatialBackdropSource();
+    auto oldSpatialBackdropSource = m_cachedSpatialBackdropSource;
```
I think it'd be a little nicer to early return before the assignment:
```
if (m_cachedSpatialBackdropSource == newSpatialBackdropSource)
    return;
m_cachedSpatialBackdropSource = newSpatialBackdropSource;
```

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1330,6 +1330,54 @@ const Color& Document::themeColor()
+        page->chrome().client().spatialBackdropSourceChanged();
+}
+
+std::optional<SpatialBackdropSource> Document::determineActiveSpatialBackdropSource()
```
I think this function can be `const`

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1330,6 +1330,54 @@ const Color& Document::themeColor()
+        if (!modelURL.isValid())
+            return std::nullopt;
+
+        return SpatialBackdropSource(sourceURL, modelURL, linkElement.environmentMap());
```
```
return SpatialBackdropSource { WTFMove(sourceURL), WTFMove(modelURL), linkElement.environmentMap() };
```

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -48,6 +48,9 @@
 #include "RenderPtr.h"
 #include "ReportingClient.h"
 #include "ScriptExecutionContext.h"
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
```
Header includes that are conditionalized like this need to go *after* regular header includes, with a blank line in between, as per coding style.

That said SpatialBackdropSource.h already has a `ENABLE(WEB_PAGE_SPATIAL_BACKDROP)` so we can just include this header unconditionally I think.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.cpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace WebCore {
+
+SpatialBackdropSource::SpatialBackdropSource(const URL& s, const URL& m, const std::optional<URL> e)
```
We can use r-value references and WTFMove() I think

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.cpp:
```diff
@@ -0,0 +1,43 @@
+
+SpatialBackdropSource::SpatialBackdropSource(const URL& sourceURL, const URL& modelURL, const URL& environmentMapURL)
+{
+    this->sourceURL = sourceURL;
```
This should use the initialization list, not the body of the constructor.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.cpp:
```diff
@@ -0,0 +1,43 @@
+SpatialBackdropSource::SpatialBackdropSource(const URL& sourceURL, const URL& modelURL, const URL& environmentMapURL)
+{
+    this->sourceURL = sourceURL;
+    this->modelURL = modelURL;
```
ditto.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.cpp:
```diff
@@ -0,0 +1,43 @@
+    this->modelURL = modelURL;
+
+    if (environmentMapURL.isValid())
+        this->environmentMapURL = environmentMapURL;
```
ditto. Can use a ternary.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+
+    WEBCORE_EXPORT SpatialBackdropSource(const URL&, const URL&, const std::optional<URL>);
+    SpatialBackdropSource(const URL& sourceURL, const URL& modelURL, const URL& environmentMapURL);
+};
```
Can use:
```
friend bool operator==(const SpatialBackdropSource&, const SpatialBackdropSource&) = default;
```

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+    SpatialBackdropSource(const URL& sourceURL, const URL& modelURL, const URL& environmentMapURL);
+};
+
+bool operator==(const SpatialBackdropSource&, const SpatialBackdropSource&);
```
Not needed with suggestion above.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+
+bool operator==(const SpatialBackdropSource&, const SpatialBackdropSource&);
+
+inline bool operator==(const SpatialBackdropSource& a, const SpatialBackdropSource& b)
```
Not needed with suggestion above.

cdumez commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -48,6 +48,9 @@
 #include "HTMLParserIdioms.h"
 #include "IdTargetObserver.h"
 #include "JSRequestPriority.h"
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
```
Should come *after* the other headers or even better, include unconditionally.

cdumez commented on Source/WebCore/html/HTMLLinkElement.cpp:
```diff
@@ -741,6 +775,18 @@ const AtomString& HTMLLinkElement::rel() const
+
+bool HTMLLinkElement::isSpatialBackdrop() const
+{
+    return m_relAttribute.isSpatialBackdrop;
```
Would be good to inline in the header since it is trivial and LinkRelAttribute.h is already included there.

cdumez commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -1070,6 +1071,7 @@ def headers_for_type(type):
         'WebCore::ShouldSample': ['<WebCore/DiagnosticLoggingClient.h>'],
         'WebCore::SourceBufferAppendMode': ['<WebCore/SourceBufferPrivate.h>'],
         'WebCore::SourceBufferEvictionData': ['<WebCore/SourceBufferPrivateClient.h>'],
+        'WebCore::SpatialBackdropSource': ['<WebCore/SpatialBackdropSource.h>'],
```
This should not be needed.

cdumez commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -576,6 +576,7 @@ def conditions_for_header(header):
         '<WebCore/ImageUtilities.h>': ["PLATFORM(COCOA)"],
         '<WebCore/MediaPlaybackTargetContext.h>': ["ENABLE(WIRELESS_PLAYBACK_TARGET)"],
         '<WebCore/PlaybackTargetClientContextIdentifier.h>': ["ENABLE(WIRELESS_PLAYBACK_TARGET)"],
+        '<WebCore/SpatialBackdropSource.h>': ["ENABLE(WEB_PAGE_SPATIAL_BACKDROP)"],
```
I doubt this is needed?

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -897,6 +897,15 @@ struct WebCore::ShareData {
 enum class WebCore::ShareDataOriginator : bool
 
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+header: <WebCore/SpatialBackdropSource.h>
```
Is this really needed? Seems like this would be the default header name based on the type name.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.mm:
```diff
@@ -0,0 +1,52 @@
+    if (!self)
+        return nil;
+
+    _sourceURL = [spatialBackdropSource->sourceURL copy];
```
Someone with better ObjC knowledge should look at this (cc @hortont424 @whsieh @ddkilzer). The `copy` looks a bit suspicious here, especially without anything in dealloc.

jeanhbr replied:
cc @achan00

achan00 replied:
I think you need a `-dealloc` to release those values.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -34,6 +34,9 @@
 #include <WebCore/NowPlayingMetadataObserver.h>
 #include <WebCore/ProcessSyncData.h>
 #include <WebCore/SnapshotIdentifier.h>
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
```
After other headers or non-conditionalized.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2820,6 +2827,9 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     void pageExtendedBackgroundColorDidChange(const WebCore::Color&);
     void sampledPageTopColorChanged(const WebCore::Color&);
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
+    void spatialBackdropSourceChanged(const std::optional<WebCore::SpatialBackdropSource>&);
```
Can use `std::optional<WebCore::SpatialBackdropSource>&&`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12508,6 +12508,27 @@ void WebPageProxy::sampledPageTopColorChanged(const Color& sampledPageTopColor)
+    return internals().spatialBackdropSource;
+}
+
+void WebPageProxy::spatialBackdropSourceChanged(const std::optional<WebCore::SpatialBackdropSource>& spatialBackdropSource)
```
Can use `std::optional<WebCore::SpatialBackdropSource>&&`

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12508,6 +12508,27 @@ void WebPageProxy::sampledPageTopColorChanged(const Color& sampledPageTopColor)
+    if (RefPtr pageClient = this->pageClient())
+        pageClient->spatialBackdropSourceWillChange();
+
+    internals().spatialBackdropSource = spatialBackdropSource;
```
`internals().spatialBackdropSource = WTFMove(spatialBackdropSource);`

cdumez commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -47,6 +47,9 @@
 #include <WebCore/PrivateClickMeasurement.h>
 #include <WebCore/RegistrableDomain.h>
 #include <WebCore/ResourceRequest.h>
+#if ENABLE(WEB_PAGE_SPATIAL_BACKDROP)
```
After other headers or conditionalized.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+    WEBCORE_EXPORT std::optional<URL> environmentMapURL;
+
+    WEBCORE_EXPORT SpatialBackdropSource(const URL&, const URL&, const std::optional<URL>);
+    SpatialBackdropSource(const URL& sourceURL, const URL& modelURL, const URL& environmentMapURL);
```
I think we can use r-value references for all parameters here and WTFMove() internally for the initialization.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+
+struct SpatialBackdropSource {
+public:
+    WEBCORE_EXPORT URL sourceURL;
```
No WEBCORE_EXPORT needed

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+struct SpatialBackdropSource {
+public:
+    WEBCORE_EXPORT URL sourceURL;
+    WEBCORE_EXPORT URL modelURL;
```
No WEBCORE_EXPORT needed

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+public:
+    WEBCORE_EXPORT URL sourceURL;
+    WEBCORE_EXPORT URL modelURL;
+    WEBCORE_EXPORT std::optional<URL> environmentMapURL;
```
No WEBCORE_EXPORT needed

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,45 @@
+    WEBCORE_EXPORT URL modelURL;
+    WEBCORE_EXPORT std::optional<URL> environmentMapURL;
+
+    WEBCORE_EXPORT SpatialBackdropSource(const URL&, const URL&, const std::optional<URL>);
```
Should all be r-value references.

cdumez reviewed (approved):
LGTM with a lot of nits and please have someone more familiar with ObjC look at the piece of code I highlighted in my review before landing.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -1330,6 +1330,47 @@ const Color& Document::themeColor()
+    auto newSpatialBackdropSource = determineActiveSpatialBackdropSource();
+    if (m_cachedSpatialBackdropSource == newSpatialBackdropSource)
+        return;
+    m_cachedSpatialBackdropSource = newSpatialBackdropSource;
```
nit: `m_cachedSpatialBackdropSource = WTFMove(newSpatialBackdropSource);`

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.cpp:
```diff
@@ -0,0 +1,37 @@
+
+namespace WebCore {
+
+SpatialBackdropSource::SpatialBackdropSource(URL&& s, URL&& m, std::optional<URL>&& e)
```
Please do not use abbreviations in WebKit. Those parameters need proper names.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.cpp:
```diff
@@ -0,0 +1,37 @@
+{
+}
+
+SpatialBackdropSource::SpatialBackdropSource(URL&& s, URL&& m, URL&& e)
```
ditto.

cdumez commented on Source/WebCore/dom/SpatialBackdropSource.h:
```diff
@@ -0,0 +1,40 @@
+    URL modelURL;
+    std::optional<URL> environmentMapURL;
+
+    WEBCORE_EXPORT SpatialBackdropSource(URL&&, URL&&, std::optional<URL>&&);
```
Omitting the parameter names here is not a good idea, especially considering they have the same time. It would be very easy for a caller to get the parameter ordering wrong.

jeanhbr committed:
Trigger a signal in the UI Process when the web page provides a spatial backdrop
https://bugs.webkit.org/show_bug.cgi?id=286510
rdar://141175983

Reviewed by Ada Chan and Chris Dumez.

Expose the value of a spatial backdrop link tag to the UI Process.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::spatialBackdropSource):
(WebCore::Document::spatialBackdropLinkElementChanged):
(WebCore::Document::determineActiveSpatialBackdropSource):
(WebCore::Document::spatialBackdropSourceChanged):
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:

* Source/WebCore/dom/SpatialBackdropSource.cpp: Added.
(WebCore::SpatialBackdropSource::SpatialBackdropSource):
* Source/WebCore/dom/SpatialBackdropSource.h: Added.
(WebCore::SpatialBackdropSource::SpatialBackdropSource):
(WebCore::operator==):
Transports the spatial backdrop properties that are useful to know for the UI
Process, this includes the source (current URL) from which the backdrop was
provided, and the URL of both the provided model and environmentmap.

* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::attributeChanged):
(WebCore::HTMLLinkElement::process):
(WebCore::HTMLLinkElement::removedFromAncestor):
(WebCore::HTMLLinkElement::isURLAttribute const):
(WebCore::HTMLLinkElement::environmentmap const):
(WebCore::HTMLLinkElement::isSpatialBackdrop const):

* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLLinkElement.idl:
* Source/WebCore/html/LinkRelAttribute.cpp:
* Source/WebCore/html/LinkRelAttribute.h:
Add the "spatial-backdrop" rel, and the "environmentmap" attribute to
the link tag.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::spatialBackdropSourceChanged const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::spatialBackdropSource const):
* Source/WebCore/page/Page.h:

* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _spatialBackdropSource]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSource.mm: Added.
(-[_WKSpatialBackdropSource initWithSpatialBackdropSource:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKSpatialBackdropSourceInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushDaemonConnection.mm:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::spatialBackdropSourceWillChange):
(WebKit::PageClientImplCocoa::spatialBackdropSourceDidChange):
* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::spatialBackdropSourceWillChange):
(WebKit::PageClient::spatialBackdropSourceDidChange):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::spatialBackdropSource const):
(WebKit::WebPageProxy::spatialBackdropSourceChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::spatialBackdropSourceChanged const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didFinishLoad):
(WebKit::WebPage::spatialBackdropSourceChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/289748@main
